Shift DTS before Kconfig

galak opened the pull request:
This is a WIP PR to get feedback on the direction and code associated with moving DTS generation before Kconfig.

The general idea is that Kconfig can utilize DTS generated values for defaults or other purposes going forward, and that there will NOT be any Kconfig based symbols or ifdef's in dts files.

To keep some functionality with dts_fixup.h files working we need to split the DTS generation into 2 pieces.  The first will preprocess the .dts file, compile dts, and generate the generated_dts_board.conf.  The second pass will run after Kconfig since we need to know CONFIG_SOC to find all the dts_fixup.h files.  The second pass will just generate the generated_dts_board.h

resolves #7302

galak committed:
dts: Remove mcuboot.overlay

As we want to remove dts dependency on Kconfig, we had a case based on
CONFIG_BOOTLOADER_MCUBOOT.  From a DTS point of view that was just
getting the chosen property 'zephyr,code-partition' set.  We can easily
move this to the actual dts files and remove the mcuboot.overlay.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: stop generating CONFIG_FLASH_LOAD_{OFFSET,SIZE}

As a step to completing removing use of Kconfig defines in dts files we
need to change how we configure/set CONFIG_FLASH_LOAD_{OFFSET,SIZE}.

Previously we would set them based on how the chosen property
'zephyr,code-partition' was set to.  If 'zephyr,code-partition' wasn't
set we would default the values to 0.  We had some generic overlay logic
which would define 'zephyr,code-partition' based on
CONFIG_BOOTLOADER_MCUBOOT.

Going forward if the DTS has 'zephyr,code-partition' set we will
generate DT_CODE_PARTITION_OFFSET & DT_CODE_PARTITION_SIZE defines.  We
will leave it to Kconfig to set CONFIG_FLASH_LOAD_OFFSET &
CONFIG_FLASH_LOAD_SIZE.

We introduce a python script that allows Kconfig to extract data from
the DTS and thus we can utilize DT_CODE_PARTITION_OFFSET and
DT_CODE_PARTITION_SIZE values to set defaults for
CONFIG_FLASH_LOAD_OFFSET & CONFIG_FLASH_LOAD_SIZE.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>
Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

galak committed:
cmake: export PROJECT_BINARY_DIR variable to kconfig

Expose PROJECT_BINARY_DIR to kconfig since we will we looking for
generated files from dts pass in the future and need to know the
location of those files.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
dts: Move CONFIG_SRAM_* & CONFIG_FLASH_* back to Kconfig

dts will now generate DT_SRAM_BASE_ADDRESS, DT_SRAM_SIZE,
DT_FLASH_BASE_ADDRESS, and DT_FLASH_SIZE defines.  Kconfig can utilize
these defines to set defaults for the CONFIG_ variants.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
boards: colibri_imx7d_m4/warp7_m4: Remove CONFIG_XIP for dts

Move the how enabling of CONFIG_XIP impacts CONFIG_FLASH_SIZE and
CONFIG_FLASH_BASE_ADDRESS to Kconfig instead of dts.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar reviewed (dismissed):
I tried to build test this, but it doesn't seem to have a solution for our use case for selecting nodes in an overlay based on kconfig:

https://github.com/foundriesio/dm-hawkbit-mqtt/blob/master/common.overlay

Is there an equivalent in DT i'm missing?

galak committed:
boards: mimxrt{1020,1050,1060,1064}_evk remove CONFIG_ from dts files

The dts files on these boards had some CONFIG_ defines related to which
memory should be used to hold code.  We move this choice out of DTS and
back into Kconfig.

As such, we removed the default setting of 'zephyr,flash' and just
map

CONFIG_CODE_ITCM to:
	DT_NXP_IMX_RT_ITCM_0_SIZE
	DT_NXP_IMX_RT_ITCM_0_BASE_ADDRESS

CONFIG_CODE_{QSPI,HYPERFLASH} to:
	DT_NXP_IMX_FLEXSPI_402A8000_SIZE_1
	DT_NXP_IMX_FLEXSPI_402A8000_BASE_ADDDRESS_1

for the mimxrt1050_evk, we remove the default setting of 'zephyr,sram'
and just map:

CONFIG_DATA_DTCM to:
	DT_NXP_IMX_DTCM_0_SIZE
	DT_NXP_IMX_DTCM_0_BASE_ADDRESS

CONFIG_DATA_SDRAM to:
	DT_MMIO_SRAM_80000000_SIZE
	DT_MMIO_SRAM_80000000_BASE_ADDRESS

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak commented:
> I tried to build test this, but it doesn't seem to have a solution for our use case for selecting nodes in an overlay based on kconfig:
> 
> https://github.com/foundriesio/dm-hawkbit-mqtt/blob/master/common.overlay
> 
> Is there an equivalent in DT i'm missing?

I'm not aware of any way, need to look into how true DTS overlay work more.  For us to be able to truly have DTS not require Kconfig, what you are doing based on SoC and CONFIG_ defines isn't something we can support.

I do think it would be interesting and useful to expand the DTC tooling to allow for conditional nodes based on some property being set.

carlescufi commented:
When building standard "hello_world" on Windows I get:

```
C:\Users\Carles\src\zephyr\samples\hello_world\b>ninja
[65/138] Building C object zephyr/CMakeFiles/zephyr.dir/soc/arm/nordic_nrf/nrf52/mpu_regions.c.obj
In file included from C:/Users/Carles/src/zephyr/soc/arm/nordic_nrf/nrf52/mpu_regions.c:8:0:
C:/Users/Carles/src/zephyr/samples/hello_world/b/zephyr/include/generated/autoconf.h:6:34: warning: large integer implicitly truncated to unsigned type [-Woverflow]
 #define CONFIG_SRAM_BASE_ADDRESS 0x536870912
                                  ^
../../../include/arch/arm/cortex_m/mpu/arm_mpu.h:44:11: note: in definition of macro 'MPU_REGION_ENTRY'
   .base = _base, \
           ^~~~~
C:/Users/Carles/src/zephyr/soc/arm/nordic_nrf/nrf52/mpu_regions.c:28:5: note: in expansion of macro 'CONFIG_SRAM_BASE_ADDRESS'
     CONFIG_SRAM_BASE_ADDRESS,
     ^~~~~~~~~~~~~~~~~~~~~~~~
[133/138] Linking C executable zephyr\zephyr_prebuilt.elf
FAILED: zephyr/zephyr_prebuilt.elf
cmd.exe /C "cd . && C:\gccarmemb\bin\arm-none-eabi-gcc.exe    zephyr/CMakeFiles/zephyr_prebuilt.dir/misc/empty_file.c.obj  -o zephyr\zephyr_prebuilt.elf  -T zephyr/linker.cmd -Wl,-Map=C:/Users/Carles/src/zephyr/samples/hello_world/b/zephyr/zephyr.map -u_OffsetAbsSyms -u_ConfigAbsSyms -Wl,--whole-archive app/libapp.a zephyr/libzephyr.a zephyr/arch/arm/core/libarch__arm__core.a zephyr/arch/arm/core/cortex_m/libarch__arm__core__cortex_m.a zephyr/arch/arm/core/cortex_m/mpu/libarch__arm__core__cortex_m__mpu.a zephyr/lib/libc/minimal/liblib__libc__minimal.a zephyr/boards/arm/nrf52840_pca10059/libboards__arm__nrf52840_pca10059.a zephyr/subsys/bluetooth/common/libsubsys__bluetooth__common.a zephyr/subsys/bluetooth/host/libsubsys__bluetooth__host.a zephyr/subsys/bluetooth/controller/libsubsys__bluetooth__controller.a zephyr/subsys/net/libsubsys__net.a zephyr/drivers/gpio/libdrivers__gpio.a zephyr/drivers/serial/libdrivers__serial.a zephyr/drivers/entropy/libdrivers__entropy.a -Wl,--no-whole-archive zephyr/kernel/libkernel.a zephyr/CMakeFiles/offsets.dir/arch/arm/core/offsets/offsets.c.obj -L"c:/gccarmemb/bin/../lib/gcc/arm-none-eabi/7.2.1/thumb/v7e-m" -LC:/Users/Carles/src/zephyr/samples/hello_world/b/zephyr -lgcc -Wl,--print-memory-usage -mthumb -mcpu=cortex-m4 -nostdlib -static -no-pie -Wl,-X -Wl,-N -Wl,--gc-sections -Wl,--build-id=none -Wl,--orphan-handling=warn -mabi=aapcs -march=armv7e-m && cd ."
Memory region         Used Size  Region Size  %age Used
           FLASH:       47476 B         1 MB      4.53%
            SRAM:       11177 B       256 KB      4.26%
        IDT_LIST:         137 B         2 KB      6.69c:/gccarmemb/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld.exe: zephyr\zephyr_prebuilt.elf section `bss' will not fit in region `SRAM'
c:/gccarmemb/bin/../lib/gcc/arm-none-eabi/7.2.1/../../../../arm-none-eabi/bin/ld.exe: region `SRAM' overflowed by 4294716329 bytes
collect2.exe: error: ld returned 1 exit status
%
ninja: build stopped: subcommand failed.
```

carlescufi commented:
Works fine on Windows now

mbolivar commented:
> I do think it would be interesting and useful to expand the DTC tooling to allow for conditional nodes based on some property being set.

Is the ELCE discussion still plan of record?

https://github.com/zephyrproject-rtos/zephyr/issues/10821

In particular these two:

```
- Build system needs to look by default for board, arch and soc .conf and .overlay files
- Find the exact arch and soc name to locate SoC-specific overlays, requires a 2-pass DT approach: First pass to find out arch and soc, second pass is actually DT (PR #9925)
```

We could definitely work with that.

It would be nice to stop cluttering the top level directory with $board.conf and $board.overlay files, by the way.

Picking these up automatically for example would be nice:

- `overlays/arch-$ARCH.overlay`
- `overlays/soc-family-$SOC_FAMILY.overlay` (if applicable)
- `overlays/soc-series-$SOC_SERIES.overlay` (if applicable)
- `overlays/soc-$SOC.overlay`
- `overlays/board-$BOARD.overlay`

That of course would require all of those values to be determined before running DTS, or during the first run of DTS, as discussed.

SebastianBoe commented:
> The second pass will run after Kconfig since we need to know CONFIG_SOC to find all the dts_fixup.h files.

Can we deduce the SOC from the BOARD in a similar manner to how we deduce the ARCH from the BOARD?

I'm concerned that a two-pass DT approach will be difficult to reason about.

galak commented on arch/arc/Kconfig:
```diff
@@ -152,6 +152,38 @@ source "arch/arc/core/mpu/Kconfig"
 
 endmenu
```
I wasn't sure if it made sense or not, but I'll move it to common arch/Kconfig.

mnkp commented on dts/common/mcuboot.overlay:
```diff
@@ -1,10 +0,0 @@
-/*
```
Rather than removing mcuboot.overlay shouldn't we move it to all the sample applications which demonstrate MCUboot usage? If this commit is accepted the `zephyr,code-partition = &slot0_partition;` chosen property will always be defined even if we don't want to place the code at slot0_partition. There is also the following comment in each board dts file:
```
		/*
		 * If chosen's zephyr,code-partion is unset, the image will be
		 * linked into the entire flash device.  If it points to an
		 * individual partition, the code will be linked to, and
		 * restricted to that partition.
		 */
```

galak replied:
I only found the comment in one file.  I've removed it now.  As for doing this in the application, I'm not against that, but would leave that to a cleanup after this PR gets merged.

galak replied:
Should be addressed.

mnkp replied:
Sorry for coming back late on this. The text is not always identical, however any board .dts file which defines `partitions` node will typically have something. E.g.
```
    /*
     * For more information, see:
     * http://docs.zephyrproject.org/latest/devices/dts/flash_partitions.html
     */
```

galak replied:
Right, do you feel we need to update the flash_partition docs?

mnkp replied:
Yes, I believe we should update the flash_partitions.html document. As for the comment within the board .dts files we can either:
- ensure that all comments refer to flash_partitions.html.
- remove all comments completely. Make sure that user is able to easily find flash_partitions.html document when searching/reading about MCUboot / bootloader support.

Also the help text of BOOTLOADER_MCUBOOT Kconfig option in ./Kconfig.zephyr needs to be updated. It currently states
> - Including dts/common/mcuboot.overlay when building the Device Tree in order to place and link the image at the slot0 offset

mnkp replied:
@galak This PR is about to be merged. I just want to remind the the documentation update related to the modified way we link flash partitions doesn't seem to be done yet. Specifically the following paragraph

If the ``chosen`` node has no ``zephyr,code-partition`` property, the
application image link uses the entire flash device. If a
``zephyr,code-partition`` property is defined, the application link
will be restricted to that partition.

from `doc/devices/dts/flash_partitions.rst` file. Unless we do it later, after release freeze.

galak replied:
Do you mind opening an issue for this, I don't want to hold up merging this on a doc update.  I'm happy to make the doc changes, just want to get this merged so it starts getting usage.

mnkp replied:
OK, will do.

nashif commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,67 @@
+
+# Types we support
+# 'string', 'int', 'hex', 'bool'
+conf_file = os.path.join(os.environ['PROJECT_BINARY_DIR'], "include/generated/generated_dts_board.conf")
```
assumption is that this plugin will be used for other things, not only DTS, so your globals here need to be namespaced and be specific to DTS and not as generic as they are now.

galak replied:
Should be fixed.

nashif commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,67 @@
+    if unit == 'g' or unit == 'g':
+        return 30
+
+def test_function_int(kconf, name, arg, unit=None):
```
please document any new macros here and in the Kconfig documentation under doc/

erwango commented:
I don't know where we stand on:
>one idea is if we match the board dir layout to soc dir so boards/arm/nrf52_pca10040 would become boards/arm/nordic_nrf/nrf52/nrf52_pca10040
-- @galak

but I don't think it makes real sense. We already have SoC factorization under dts, doing similar factorization under board is redundant and hence should not provide much information
If factorization should happen under board, it should rather be form factor/bsp related, for instance

galak committed:
cmake: Run dts before kconfig

Based on work from Sebastian Bøe and updated to current tree.

Signed-off-by: Sebastian Bøe <sebastian.boe@nordicsemi.no>
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

dbkinder commented:
Some doc changes coming too?

galak committed:
doc: Add docs about kconfigfunctions

Add docs on the kconfigfunctions that we support.  We only expose some
functions related to device tree currently.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

SebastianBoe commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,62 @@
+        d = d >> _dt_units_to_scale(unit)
+        d = hex(d)
+
+    except:
```
It is not OK to catch "all" exceptions, as that can lead to impossibly difficult debugging when an unexpected exception occurs, like OOM / syntax error, etc.

galak commented:
> Some doc changes coming too?

@dbkinder I introduced `doc/porting/kconfig.rst` can you take a look and help me with it.

SebastianBoe reviewed (changes requested):
I'm sceptical that it is necessary to run DT twice and would like to see it reduced to one pass to limit complexity and configure-time.

Could it be explained in detail why one pass is troublesome so that this can be peer-reviewed properly?

nashif commented on doc/porting/kconfig.rst:
```diff
@@ -0,0 +1,28 @@
+.. _kconfig:
```
this should go into doc/tools/kconfig/..

galak replied:
Moved it there.

galak commented:
> Could it be explained in detail why one pass is troublesome so that this can be peer-reviewed properly?

We are generating a few different files during the various "passes".  During the first pass we generate the `<BOARD>.dts.pre.tmp`, `<BOARD>.dts_compiled`, and `include/generated/generated_dts_board.conf`.

The second "pass" generates `include/generated/generated_dts_board.h`.  The reason is that the 'include/generated/generated_dts_board.h' includes 'the dts_fixup.h' and we need to know CONFIG_SOC and the various SOC path's to find the dts_fixup.h files.  We don't know the SOC path's until Kconfig runs.  Thus we have the pass before Kconfig and the pass after.

dbkinder commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,28 @@
+.. _functions:
```
label here should probably be more specific:  ``kconfig-functions`` (to avoid future label conflicts)

dbkinder commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,28 @@
+Kconfig Functions
+#################
+
+The Kconfiglib provides a means for user defined preprocessor functions.  We
```
We could simplify this to something like this:
```
Kconfiglib provides user-defined preprocessor functions that
we use in Zephyr to expose Device Tree information to Kconfig.
For example, we can get the default value for a Kconfig symbol
from the device tree.
```

dbkinder commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,28 @@
+Device Tree Related Functions
+*****************************
+
+dt_get_int_val:
```
Should you add the parameters to these functions?  Something like:
```
dt_get_int_val (name, unit)
```

galak replied:
Added

galak commented:
@dbkinder made the suggested changes.

SebastianBoe commented:
Aha, when the word "pass" was used, I assumed we were running DTC multiple times. But it is extract_dts_includes.py that is being run multiple times.

But still, we don't need python to concatenate C header files. The C preprocessor should do just fine.

To reduce the amount of python code generation, to reduce the complexity of extract_dts_includes.py, and to increase the use of the C language I would suggest that we do the following:

- have extract_dts_includes.py generate an **un**fixed header file.
- have generated_dts_board.h be a source file instead of a generated file
- have generated_dts_board.h include the unfixed header file and any fixup header files.

This would look something like this:

https://github.com/zephyrproject-rtos/zephyr/compare/master...SebastianBoe:simplify_fixup_mechanism?expand=1#diff-ce337d7116685c9c894ff4a4108c5dc7

SebastianBoe commented on cmake/kconfig.cmake:
```diff
@@ -30,6 +30,7 @@ set(ENV{PYTHON_EXECUTABLE} ${PYTHON_EXECUTABLE})
 set(ENV{ARCH}      ${ARCH})
 set(ENV{BOARD_DIR} ${BOARD_DIR})
 set(ENV{SOC_DIR}   ${SOC_DIR})
+set(ENV{PROJECT_BINARY_DIR} ${PROJECT_BINARY_DIR})
```
Yes. This will set it for any processes created during Configure-time only, the below will set it for the process that is created at build time.

galak replied:
It appears to be required, removing causes nothing to work :)

ulfalizer replied:
Alright... as long as it's needed both configure- and build time I guess.

galak commented:
> Aha, when the word "pass" was used, I assumed we were running DTC multiple times. But it is extract_dts_includes.py that is being run multiple times.
> 
> But still, we don't need python to concatenate C header files. The C preprocessor should do just fine.
> 
> To reduce the amount of python code generation, to reduce the complexity of extract_dts_includes.py, and to increase the use of the C language I would suggest that we do the following:
> 
> * have extract_dts_includes.py generate an **un**fixed header file.
> * have generated_dts_board.h be a source file instead of a generated file
> * have generated_dts_board.h include the unfixed header file and any fixup header files.
> 
> This would look something like this:
> 
> [zephyrproject-rtos/zephyr/compare/master...SebastianBoe:simplify_fixup_mechanism?expand=1#diff-ce337d7116685c9c894ff4a4108c5dc7](https://github.com/zephyrproject-rtos/zephyr/compare/master...SebastianBoe:simplify_fixup_mechanism?expand=1#diff-ce337d7116685c9c894ff4a4108c5dc7)

Works for me, I'm all for having the scripts do less.  Can you work this up as a proper PR?

galak commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,30 @@
+dt_get_int_val(kconf, name, arg):
+    This function looks up 'name' in the DTS generated "conf" style database
+    and if its found it will return the value as an decimal integer.  The
+    function will scale the value based on 'unit':
```
Fixed.

galak commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,30 @@
+Device Tree Related Functions
+*****************************
+
+dt_get_int_val(kconf, name, arg):
```
I'm all for shorter names, will drop the `get_`

galak commented on arch/Kconfig:
```diff
@@ -80,6 +80,42 @@ config BIG_ENDIAN
+
+config SRAM_SIZE
+	int "SRAM Size in kB"
+	default "$(dt_get_int_val,DT_SRAM_SIZE)"
```
Fixed

galak commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,30 @@
+*****************************
+
+dt_get_int_val(kconf, name, arg):
+    This function looks up 'name' in the DTS generated "conf" style database
```
Added an example to the doc.

galak commented:
@ulfalizer addressed your comments.

SebastianBoe commented:
> Works for me, I'm all for having the scripts do less. Can you work this up as a proper PR?

Will do.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+    """
+    try:
+        d = dt_defines[arg]
+        if d.startswith('0x') or d.startswith('0x'):
```
Typo - `0x` twice.

`startswith()` and `endswith()` have a handy feature for testing many values at once by the way. Can pass a tuple of values to test like this:

```python
if d.startswith(('0x', '0X')):
    ...
```

galak replied:
Fixed & Changed.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+            d = int(d, 16)
+        else:
+            d = int(d)
+        d = d >> _dt_units_to_scale(unit)
```
Nit: There's `>>=` in Python

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+        d = d >> _dt_units_to_scale(unit)
+
+    except KeyError:
+        d = 0
```
Another option would be to error out in some way here I guess:

```python
sys.exit("{}:{}: unknown unit '{}'".format(kconf._filename, kconf._linenr))
```

That's using some internal variables to get the location, but it should be pretty safe.

galak replied:
The sys.exit() might not work, is I think we need a default value for the cases in which the key doesn't exist.

For example if we build for posix since it doesn't have dt we need to default value and not error out in this case.

ulfalizer replied:
Ah, didn't notice that the `try` also surrounds the `d = dt_defines[arg]`.

Maybe the code could do this instead (with no `try`), to make it a bit more explicit:

```python
if name not in dt_defines:
    return "0"

...
```

One thing that often confuses me in the `scripts/dts` code is that there are a lot of  `try` blocks around a bunch of code where it's not obvious what it means when you catch the `KeyError`, etc.

galak replied:
That's from my lack of know going python style ;)

galak replied:
Changed to do the `if` check and removed the try block.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+        d = hex(d)
+
+    except KeyError:
+        d = 0x0
```
If this is turned into an error instead, then maybe the `sys.exit(...)` (see above) could be moved into `_dt_units_to_scale()` instead.

ulfalizer replied:
(Could ignore this. Was confused about what the `KeyError` was about.)

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+    """
+    try:
+        d = dt_defines[arg]
+        if d.startswith('0x') or d.startswith('0x'):
```
Ditto as in `dt_int_val()`.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,78 @@
+dt_defines = {}
+if os.path.isfile(conf_file):
+    for line in open(conf_file, 'r', encoding='utf-8'):
+        if re.search('=', line):
```
Can do `if '=' in line:` instead (and get rid of the `import re`).

galak replied:
Done.

ulfalizer reviewed (changes requested):
Forgot to look at `kconfigfunctions.py`. Sorry about that.

The other fixes look fine.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,77 @@
+
+def dt_int_val(kconf, name, arg, unit=None):
+    """
+    This function looks up 'name' in the DTS generated "conf" style database
```
`name` is the name of the function (`"dt_int_val"`, think `argv[0]`), so the documentation is wrong.

Maybe the function signature could be something like this instead, since the name of the function isn't needed:

```python
def dt_int_val(kconf, _, name, unit=None):
    ...
```

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,77 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import sys
```
Unused import?

galak replied:
Looks so, removed.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,77 @@
+    if unit == 'm' or unit == 'M':
+        return 20
+    if unit == 'g' or unit == 'G':
+        return 30
```
Just a comment, but can do `if unit in {'k', 'K'}:`, etc., too.

galak replied:
Done

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,77 @@
+    d >>= _dt_units_to_scale(unit)
+    d = hex(d)
+
+    return str(d)
```
Could get rid of the `str()`. `hex()` returns a string already.

galak replied:
Done

ulfalizer reviewed (approved):
Just minor nits, and need to do some panicky apartment stuff, so approved.

galak commented:
@dbkinder ping

mnkp reviewed (commented):
What is the rationale behind moving `FLASH_SIZE`, `FLASH_BASE_ADDRESS` back to Kconfig? Is it because we removed mcuboot.overlay which was conditionally included based on Kconfig option and now we need a way to support flash partitions for MCUboot?

galak commented:
> What is the rationale behind moving `FLASH_SIZE`, `FLASH_BASE_ADDRESS` back to Kconfig? Is it because we removed mcuboot.overlay which was conditionally included based on Kconfig option and now we need a way to support flash partitions for MCUboot?

The short answer is yes, the long answer is its to keep DTS free of Kconfig symbols and still provide a means for the user to pick.

galak commented on Kconfig.zephyr:
```diff
@@ -82,9 +82,9 @@ config HAS_FLASH_LOAD_OFFSET
-if !HAS_DTS
 config FLASH_LOAD_OFFSET
 	hex "Kernel load offset"
+	default $(dt_hex_val,DT_CODE_PARTITION_OFFSET) if BOOTLOADER_MCUBOOT
```
While that's reasonable, I don't want to over burden this PR with cleanups that can occur after the fact.  The goal is to maintain the functionality we have today.

galak commented:
@ulfalizer need help with how to update doc/CMakefile.txt for my kconfigfunction

mnkp commented:
If removing the DTS dependency on `BOOTLOADER_MCUBOOT` Kconfig symbol was the only reason to move `FLASH_SIZE`, `FLASH_BASE_ADDRESS` back to Kconfig then maybe there is an easier way to do it. I already suggested it but I'm not sure if I was clear enough. We could leave the old setup in place and move mcuboot.overlay from dts/common/ directly to MCUBoot sample application. And remove from dts/common/common.dts
```
#ifdef CONFIG_BOOTLOADER_MCUBOOT
#include "mcuboot.overlay"
#endif
```
This way the DTS dependency on `BOOTLOADER_MCUBOOT` Kconfig symbol is gone.

There are some strings attached: 
- We would need to add support for an .overlay file which is not board specific. I believe this is missing.
- This approach would need to be aligned with MCUBoot project as they provide sample application.

The current approach where linking is driven exclusively by `zephyr,code-partition` property in DTS `chosen` node is quite convenient. Changing it seems like a lot of work and the new setup is not necessary more straightforward to use.

galak commented:
> If removing the DTS dependency on `BOOTLOADER_MCUBOOT` Kconfig symbol was the only reason to move `FLASH_SIZE`, `FLASH_BASE_ADDRESS` back to Kconfig then maybe there is an easier way to do it. I already suggested it but I'm not sure if I was clear enough. We could leave the old setup in place and move mcuboot.overlay from dts/common/ directly to MCUBoot sample application. And remove from dts/common/common.dts
> 
> ```
> #ifdef CONFIG_BOOTLOADER_MCUBOOT
> #include "mcuboot.overlay"
> #endif
> ```
> This way the DTS dependency on `BOOTLOADER_MCUBOOT` Kconfig symbol is gone.
> 
> There are some strings attached:
> 
> * We would need to add support for an .overlay file which is not board specific. I believe this is missing.
> * This approach would need to be aligned with MCUBoot project as they provide sample application.
> 
> The current approach where linking is driven exclusively by `zephyr,code-partition` property in DTS `chosen` node is quite convenient. Changing it seems like a lot of work and the new setup is not necessary more straightforward to use.

Mcuboot isn't the only case, on the i.mx-rt we allow selecting different flash or memory spaces to boot from.  There are examples like XIP or not that impact things.

galak commented:
> * We would need to add support for an .overlay file which is not board specific. I believe this is missing.

How would you select this overlay?  Can't be existing Kconfig, so now you have some other means that user has to be aware of.  It starts getting complicated pretty quickly.

mnkp commented:
> Mcuboot isn't the only case, on the i.mx-rt we allow selecting different flash or memory spaces to boot from. There are examples like XIP or not that impact things.

Clear, that was just a suggestion.

> How would you select this overlay? Can't be existing Kconfig, so now you have some other means that user has to be aware of. It starts getting complicated pretty quickly.

I was talking about an overlay file which is always sourced if present in the application directory. E.g. default.overlay, app.overlay, etc. So far it looks like we don't need one.

SebastianBoe commented on dts/common/common.dts:
```diff
@@ -2,7 +2,3 @@
  * Common Device Tree source, used for conditionally pulling in features and
```
It is no longer possible to conditionally pull in features. So we shouldn't have a file saying
that we can do so.

galak replied:
Fixed.

SebastianBoe commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,75 @@
+
+dt_defines = {}
+if os.path.isfile(conf_file):
+    for line in open(conf_file, 'r', encoding='utf-8'):
```
'with' must be used.

AFAICT this code has a resource leak[0].

[0] https://stackoverflow.com/a/25070939

galak replied:
Changed to use 'with'

SebastianBoe commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,53 @@
+dt_int_val(kconf, _, name, unit):
+    This function looks up 'name' in the DTS generated "conf" style database
+    and if its found it will return the value as an decimal integer.  The
+    function will scale the value based on 'unit':
```
Could you use either the word multiply or divide instead of scale, I'm not able to tell in which direction it scales the value.

galak replied:
Changed to divide or division.

galak commented on doc/tools/kconfig/functions.rst:
```diff
@@ -0,0 +1,53 @@
+In this example if we examine the generated generated_dts_board.conf file
+as part of the Zephyr build we'd find the following entry:
+
+   DT_NXP_IMX_FLEXSPI_402A8000_SIZE_1=8388608
```
8388608 is the correct value.

galak commented:
@dbkinder can you review the doc update/changes.

mnkp commented on boards/arm/mimxrt1050_evk/mimxrt1050_evk.dts:
```diff
@@ -28,11 +28,6 @@
-#if defined(CONFIG_CODE_ITCM)
-		zephyr,flash = &itcm0;
-#elif defined(CONFIG_CODE_HYPERFLASH)
-		zephyr,flash = &hyperflash0;
```
Wouldn't it be cleaner if the selection was handled by a board .overlay file? E.g. `mimxrt1050_evk_itcm.overlay`, `mimxrt1050_evk_hyperflash.overlay`. We would keep usage of `zephyr,flash` property consistent and FLASH_BASE_ADDRESS could still be defined in DTS. The selection of board overlay file could be done in CMakeLists.txt based on Kconfig option.

galak replied:
The problem than is you've made it so that the full dts isn't know until after Kconfig runs, which means we can't utilize the .dts in things like sanitycheck or Kconfig itself.

mnkp replied:
Of course, sorry. That's not going to work.

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,80 @@
+if os.path.isfile(conf_file):
+    with open(conf_file, 'r', encoding='utf-8') as fd:
+        line = fd.readline()
+        while line:
```
Could do this:

```python
with open(...) as fd:
    for line in fd:
        ...
```

(And remove the final `readline()` too.)

galak replied:
Done

ulfalizer commented on scripts/kconfig/kconfigfunctions.py:
```diff
@@ -0,0 +1,80 @@
+        line = fd.readline()
+        while line:
+            if '=' in line:
+                c = line.split('=')
```
Another option is `key, val = line.split('=')`.

galak replied:
not sure I follow how this change would work.

ulfalizer replied:
`key, val = 'foo=bar'.split('=')` sets `key` to `foo` and `val` to `bar`. Python "unpacks" the RHS when you have multiple variables in the LHS.

Would make the code afterwards `dt_defines[key] = val.strip()`.

ulfalizer replied:
Would get this along with the other change:

```python
dt_defines = {}
if os.path.isfile(conf_file):
    with open(conf_file, 'r', encoding='utf-8') as fd:
        for line in fd:
            if '=' in line:
                key, val = line.split('=')
                dt_defines[key] = val.strip()
```

I wonder if the `strip()` is needed.

galak replied:
The strip() is needed, made this change.

ulfalizer reviewed (approved):
Just some small suggestions.

carlescufi committed:
doc: Add KCONFIG_DOC_MODE env var handling

Export a new KCONFIG_DOC_MODE environment variable when building the doc
and invoking Kconfig, so that the functions that expect a build folder
can accordingly return a hardcoded value.

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

galak commented:
@dbkinder can you review the docs again?

galak commented:
@dbkinder can you look at the docs again?

mbolivar commented:
@galak I've submitted a PR for the west commands which hopefully should make it clear why CONFIG_BOOTLOADER_MCUBOOT is a big usability win:

https://github.com/zephyrproject-rtos/zephyr/pull/12903

I would like to work with you to make sure we can keep the usability of building, signing, and flashing MCUboot child images in Zephyr even if DTS comes before Kconfig. Do you have time for that this week/next?

Thanks!

SebastianBoe commented on arch/Kconfig:
```diff
@@ -80,6 +80,42 @@ config BIG_ENDIAN
+if ARC || ARM || NIOS2 || X86
+
+config SRAM_SIZE
+	int "SRAM Size in kB"
```
I assume that this is unrelated to the PR, so no requirement to reply, but for the record;

I believe we should generally use bytes everywhere, and only translate to other units when necessary (e.g. when presenting human readable data).

SebastianBoe replied:
Also, when not using the "standard" (context-dependent) unit, one should include the unit information in the symbol;

so SRAM_SIZE_IN_KB.

But I assume that CONFIG_SRAM_SIZE is a pre-existing concept in some form and that changing this is out-of-scope for this PR.

galak replied:
Agreed, its needs cleaning up across the code base.

galak replied:
Yes, that change is outside of the scope of this PR.

SebastianBoe commented:
@galak : I believe you've posted a PR somewhere that runs DTC multiple times, is that still on the table?

Or does this replace that PR?

galak commented:
> @galak : I believe you've posted a PR somewhere that runs DTC multiple times, is that still on the table?
> 
> Or does this replace that PR?

I think that's still on the table but for different reasons.
